### PR TITLE
tsc commands in packages dir use --build mode (incremental builds)

### PR DESCRIPTION
**Pull request by mreishus**
#### Changes proposed in this Pull Request

* `tsc` commands in the `packages/` directory use `--build` mode, which enables [incremental builds](https://www.typescriptlang.org/docs/handbook/project-references.html).
  * `yarn start`, then control-c, then `yarn start` again won't rebuild all of the packages, because `tsc` in build mode uses [up-to-date checking](https://www.typescriptlang.org/docs/handbook/project-references.html).
* I don't know a lot about our mono-repo, so it's possible this breaks something.  However, simply using calypso seems to work fine for me.

##### Establish Baseline Timings
* `git checkout master`
* `find . -name 'node_modules' -type d -prune -exec rm -rf '{}' +`
* `yarn && yarn start`  _(note time taken to see "wp-calypso booted")_
* Control-C
* `yarn start` _(note time taken to see "wp-calypso booted")_

##### Try this branch
* `git checkout try/typescript-incremental-compilation`
* `find . -name 'node_modules' -type d -prune -exec rm -rf '{}' +`
* `yarn && yarn start`  _(note time taken to see "wp-calypso booted")_
* Control-C
* `yarn start` _(note time taken to see "wp-calypso booted")_

### CODE REVIEW

**Comment by mreishus** [comment]
My timings

```
=====================
-- MASTER BRANCH
=====================
> Delete all node modules
> yarn && yarn start
--> took 3:30 to see "wp-calypso booted"

> Control-C
> yarn start
---> took 1:03 to see "wp-calypso booted"

> Control-C
> yarn start
---> Took 1:04 to see "wp-calypso booted"

=====================
-- THIS BRANCH
=====================
> Delete all node modules
> yarn && yarn start
---> took 1:50 to see "wp-calypso booted"

> Control-C
> yarn start
---> took 0:24 to see "wp-calypso booted"

> Control-C
> yarn start
---> Took 0:25 to see "wp-calypso booted"
```

**Review by sirbrillig — APPROVED** [review]
I didn't follow the instructions quite right; I waited for "Ready! You can load..." instead of "wp-calypso booted" but I think my results are probably still valuable since it seems that for me "Ready" always comes just about a minute after "wp-calypso booted" and I was at least consistent. Note that I have a 2 core 2017 MacBook Pro.

###### master branch

- Deleted all `node_modules`
- Took 7:30 from `yarn && yarn start` to see `Ready!`
- Ctrl-C
- Took 3:00 from `yarn start` to see `Ready!`

###### try/typescript-incremental-compilation branch

- Deleted all `node_modules`
- Took 4:30 from `yarn && yarn start` to see `Ready!`
- Ctrl-C
- Took 1:40 from `yarn start` to see `Ready!`

Looks great to me 🎉 , so I'll approve but I think you should also wait for a review from someone in Team Calypso who knows something about tsc because I'm just a TS amateur.

**Comment by scinos** [comment]
Looks great! 
Pinging our resident TS expert, @sirreal

**Comment by scinos** [comment]
Some data from my local machine:

Script:
```
# Clean TSC cache
# Clean TSC compilation results
# Compile packages (like `yarn start` does) with cold cache and then warm cache
rm -fr .tsc-cache && rm -fr packages/**/dist && yarn run build-packages && yarn run build-packages
```

Results (avg of 3 runs):

|             | cold cache | warm cache |
|-------------|------------|------------|
| master      | 79.74s     | 52.46s     |
| This branch | 87.4s      | **9.06s** ⭐  |


`

**Review by sirreal — APPROVED** [review]
👍  This looks good to me. I think it's a step in the right direction, thank you!

I'll take the opportunity to highlight improvements we can still make. References give us the ability to only build whats needed, on demand, and only when it needs to be rebuilt. There's still a fair amount of potential redundancy right now, `lerna` will spin up a bunch of processes in parallel that will build things redundantly.

A good intermediate step might be something like the following. We add a single `tsconfig` that just references all the projects built by tsc, e.g:

```jsonc
// packages/tsconfig.json
{
  "files": [],
  "references": [
    { "path": "data-stores" },
    // All tsc-built packages
  ],
}
```

Then, if we run `tsc --build packages` and `tsc --build packages --watch` for dev mode along with existing webpack processes.

This seems pretty simple and largely aligns with what we do with lerna right now.

A more ideal setup might be to integrate a loader that understands project references, e.g. [`ts-loader`](https://www.npmjs.com/package/ts-loader) with its [support for references](https://github.com/TypeStrong/ts-loader/blob/HEAD/REFERENCES.md).

**Comment by scinos** [comment]
> This seems pretty simple and largely aligns with what we do with lerna right now.

+1 to this idea. In fact, if we do that, we can get rid of lerna entirely. AFAIK we only use to run `prepare` in all TS projects.

**Review by tyxla — APPROVED** [review]
Works great! Amazing work 😍  🥇

**Pull request merged**
